# 안녕하세요^^ 
## AIVLE 미니프로젝트 '서울시 생활정보 기반 대중교통 수요 분석 ' 과정에 오신 여러분을 환영합니다.
* 본 과정에서는 실제 사례와 데이터를 기반으로 문제를 해결하는 전체 과정을 자기 주도형 실습으로 진행해볼 예정입니다.
* 앞선 교육과정을 정리하는 마음과 지금까지 배운 내용을 바탕으로 문제 해결을 해볼게요!
* 미니 프로젝트를 통한 문제 해결 과정 'A에서 Z까지', 지금부터 시작합니다!

## 개인과제
* 데이터 구조, 분포 확인, 전처리 : 1.1_버스정류장데이터 / 1.2 유동인구데이터 / 1.3_주민등록데이터 / 1.4_업종등록데이터
* 데이터간 관계 가설 수립 및 검증 : 1.5_데이터 분석

---

# 데이터 분석부터 먼저 시작해보겠습니다.
## "구 별 유동인구 데이터" 를 확인해 보도록 하겠습니다

In [33]:
# 필요 라이브러리부터 설치합니다.
# %pip install pandas seaborn

import sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 기본전제

[기본 데이터]
* 1.2 seoul_moving_month_4.csv

[데이터 소개]
* 서울 시 구별 이동 2022년 4월 데이터

[변수 소개]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
    - 0세 = 0~9세, 80세 = 80~세
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리

---

# 1.데이터 불러오기
## 모든 미니 프로젝트의 시작은 '데이터 불러오기' 부터라고 할 수 있습니다.
+ KeyPoint : 불러오고자 하는 데이터에 따라 자유롭게 변수로 지정할 수 있다.

###  데이터 프레임을 불러오고 변수로 저장(여기서는 CSV 기준으로 진행)
* csv : pd.read_csv("파일이름. csv")
* txt : pd.read_csv("파일이름. csv", sep="구분자")
* xlsx : pd.read_excel('파일이름.xlsx')
* pickle : pd.read_pickle("파일이름.pkl") <br>
 [참고] pickle은 파이썬의 모든 객체를 파일로 저장할 수 있는 방법으로 DataFrame,List,Dict 등 모든 객체 저장 가능(특히 sklearn라이브러리를 통해 모델을 학습시키고, 저장할 때 많이 사용)

#### [실습문제1] 데이터 로딩
* Pandas 라이브러리를 활용해서 'seoul_moving_month_4_step1.csv' 데이터를 확인하세요.
    * 아래에 'Pandas'를 Import하고 '1.2 seoul_moving_month_4.csv'파일을 'seoul_moving'에 불러오고,  Shape을 확인하는 소스코드를 작성하고 실행해보세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요
        * index column은 0번 열 입니다

In [34]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
import pandas as pd
movement = pd.read_csv('C:/Users/User/iCloudDrive/AIVLE/3주차/2023.02.13_미니프로젝트1차_실습자료/1.2 seoul_moving_month_4.csv')
movement


Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                  0  202204  일     0      11010      11010  F   0   EE   
1                  1  202204  일     0      11010      11010  F   0   EH   
2                  2  202204  일     0      11010      11010  F   0   HE   
3                  3  202204  일     0      11010      11010  F  10   HW   
4                  4  202204  일     0      11010      11010  F  10   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28923049      968458  202204  토    23      39000      11250  M  45   EH   
28923050      968459  202204  토    23      39000      11250  M  50   EH   
28923051      968460  202204  토    23      39000      11250  M  55   EE   
28923052      968461  202204  토    23      39000      11250  M  55   EH   
28923053      968462  202204  토    23      39000      11250  M  75   EH   

          평균 이동 시간(분) 이동인구(합)  
0                  10   28.02  
1                  10   27.52  
2                  10   27.83  
3                  10    6.05  
4                  90    5.95  
...               ...     ...  
28923049           50    3.16  
28923050           70    3.27  
28923051           50    3.31  
28923052           50    3.28  
28923053           30    5.94  

[28923054 rows x 11 columns]

In [35]:
movement['요일'].value_counts()

토    4690806
금    4641881
목    4011291
수    3955646
화    3926876
월    3906272
일    3790282
Name: 요일, dtype: int64

In [36]:
# 데이터 프레임의 Shape을 확인합니다.
movement.describe()

Unnamed: 0        대상연월          도착시간     출발 시군구 코드     도착 시군구 코드  \
count  2.892305e+07  28923054.0  2.892305e+07  2.892305e+07  2.892305e+07   
mean   7.141390e+05    202204.0  1.365597e+01  1.712362e+04  1.719750e+04   
std    4.753792e+05         0.0  5.849675e+00  9.046478e+03  9.063818e+03   
min    0.000000e+00    202204.0  0.000000e+00  1.101000e+04  1.101000e+04   
25%    3.012810e+05    202204.0  1.000000e+01  1.110000e+04  1.111000e+04   
50%    6.556670e+05    202204.0  1.400000e+01  1.120000e+04  1.120000e+04   
75%    1.082991e+06    202204.0  1.800000e+01  3.101400e+04  3.102100e+04   
max    1.871922e+06    202204.0  2.300000e+01  3.900000e+04  3.900000e+04   

                 나이   평균 이동 시간(분)  
count  2.892305e+07  2.892305e+07  
mean   4.279614e+01  3.279829e+01  
std    1.777485e+01  2.584213e+01  
min    0.000000e+00  1.000000e+01  
25%    3.000000e+01  2.000000e+01  
50%    4.500000e+01  3.000000e+01  
75%    5.500000e+01  4.000000e+01  
max    8.000000e+01  7.700000e+02

In [37]:
movement.shape

(28923054, 11)

---

# 2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [실습문제2] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head' 등 전부 활용해 보겠습니다.

In [38]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
movement.describe()

Unnamed: 0        대상연월          도착시간     출발 시군구 코드     도착 시군구 코드  \
count  2.892305e+07  28923054.0  2.892305e+07  2.892305e+07  2.892305e+07   
mean   7.141390e+05    202204.0  1.365597e+01  1.712362e+04  1.719750e+04   
std    4.753792e+05         0.0  5.849675e+00  9.046478e+03  9.063818e+03   
min    0.000000e+00    202204.0  0.000000e+00  1.101000e+04  1.101000e+04   
25%    3.012810e+05    202204.0  1.000000e+01  1.110000e+04  1.111000e+04   
50%    6.556670e+05    202204.0  1.400000e+01  1.120000e+04  1.120000e+04   
75%    1.082991e+06    202204.0  1.800000e+01  3.101400e+04  3.102100e+04   
max    1.871922e+06    202204.0  2.300000e+01  3.900000e+04  3.900000e+04   

                 나이   평균 이동 시간(분)  
count  2.892305e+07  2.892305e+07  
mean   4.279614e+01  3.279829e+01  
std    1.777485e+01  2.584213e+01  
min    0.000000e+00  1.000000e+01  
25%    3.000000e+01  2.000000e+01  
50%    4.500000e+01  3.000000e+01  
75%    5.500000e+01  4.000000e+01  
max    8.000000e+01  7.700000e+02

In [39]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
movement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28923054 entries, 0 to 28923053
Data columns (total 11 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   Unnamed: 0   int64 
 1   대상연월         int64 
 2   요일           object
 3   도착시간         int64 
 4   출발 시군구 코드    int64 
 5   도착 시군구 코드    int64 
 6   성별           object
 7   나이           int64 
 8   이동유형         object
 9   평균 이동 시간(분)  int64 
 10  이동인구(합)      object
dtypes: int64(7), object(4)
memory usage: 2.4+ GB


In [40]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
movement.head()

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0           0  202204  일     0      11010      11010  F   0   EE           10   
1           1  202204  일     0      11010      11010  F   0   EH           10   
2           2  202204  일     0      11010      11010  F   0   HE           10   
3           3  202204  일     0      11010      11010  F  10   HW           10   
4           4  202204  일     0      11010      11010  F  10   EH           90   

  이동인구(합)  
0   28.02  
1   27.52  
2   27.83  
3    6.05  
4    5.95

* 1번 열의 의미
    * " 2022년 1월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 H(야간상주지(집))으로 이동한 0 ~ 10세 여성은 총 27.52명이며 이들의 평균 이동시간은 10분이다

#### [실습문제3] 결측치 처리 ( NaN이 아닌 결측치를 처리 해봅시다) << 중요 >>

* '인구이동(합)' column은 이동 인구의 합산인데 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리)

    - 해당 값을 나름의 기준으로 변경하세요

    - '인구이동(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [41]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (Hint : value_counts())
movement['이동인구(합)'].value_counts()

*          3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [42]:
# movement['이동인구합(0제외)'] = movement.loc[movement['이동인구(합)'] == 0, '이동인구(합)'] = 1.5
# movement.replace("*","1.5", inplace=True)
movement.loc[movement['이동인구(합)'] == '*'] = 1.5
movement.head()

Unnamed: 0      대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별    나이 이동유형  \
0         0.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   EE   
1         1.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   EH   
2         2.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   HE   
3         3.0  202204.0  일   0.0    11010.0    11010.0  F  10.0   HW   
4         4.0  202204.0  일   0.0    11010.0    11010.0  F  10.0   EH   

   평균 이동 시간(분) 이동인구(합)  
0         10.0   28.02  
1         10.0   27.52  
2         10.0   27.83  
3         10.0    6.05  
4         90.0    5.95

In [43]:
movement['이동인구(합)'].value_counts()

1.5        3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [44]:
movement['도착시간'].value_counts()

1.5     3842041
18.0    1659905
17.0    1633838
19.0    1572914
16.0    1552113
15.0    1490388
13.0    1477207
14.0    1468998
12.0    1428266
20.0    1346662
11.0    1328539
10.0    1236173
21.0    1168320
9.0     1147895
8.0     1035948
22.0    1019658
7.0      840710
23.0     804102
6.0      652506
0.0      616972
1.0      413843
5.0      401830
2.0      290560
4.0      256392
3.0      237274
Name: 도착시간, dtype: int64

In [45]:
# 최빈값과 평균값을 알 수 없는 상황이라 중위값으로 대체
movement.replace('*', '1.5', inplace=True)

In [46]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
movement['이동인구(합)'].value_counts()

1.5        3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [47]:
# 이동인구(합) column의 데이터 타입을 바꿔주세요
# astype
movement['이동인구(합)'] = movement['이동인구(합)'].astype('float')

#### [실습문제4] 이동유형, 요일, 성별 변수의 value수를 파악

* 역시나 value_counts()를 활용해 봅시다.

In [48]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
movement['이동유형'].value_counts()

EH     5071239
HE     4862072
EE     4377788
1.5    3842041
WH     2770340
HW     2432742
WE     2343722
EW     1875474
WW      873431
HH      474205
Name: 이동유형, dtype: int64

In [49]:
movement['요일'].value_counts()

토      4107001
금      4044366
1.5    3842041
목      3473409
수      3428786
화      3404321
월      3368429
일      3254701
Name: 요일, dtype: int64

In [50]:
movement['성별'].value_counts()

M      13970365
F      11110648
1.5     3842041
Name: 성별, dtype: int64

In [51]:
movement

Unnamed: 0      대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별    나이 이동유형  \
0                0.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   EE   
1                1.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   EH   
2                2.0  202204.0  일   0.0    11010.0    11010.0  F   0.0   HE   
3                3.0  202204.0  일   0.0    11010.0    11010.0  F  10.0   HW   
4                4.0  202204.0  일   0.0    11010.0    11010.0  F  10.0   EH   
...              ...       ... ..   ...        ...        ... ..   ...  ...   
28923049    968458.0  202204.0  토  23.0    39000.0    11250.0  M  45.0   EH   
28923050    968459.0  202204.0  토  23.0    39000.0    11250.0  M  50.0   EH   
28923051    968460.0  202204.0  토  23.0    39000.0    11250.0  M  55.0   EE   
28923052    968461.0  202204.0  토  23.0    39000.0    11250.0  M  55.0   EH   
28923053    968462.0  202204.0  토  23.0    39000.0    11250.0  M  75.0   EH   

          평균 이동 시간(분)  이동인구(합)  
0                10.0    28.02  
1                10.0    27.52  
2                10.0    27.83  
3                10.0     6.05  
4                90.0     5.95  
...               ...      ...  
28923049         50.0     3.16  
28923050         70.0     3.27  
28923051         50.0     3.31  
28923052         50.0     3.28  
28923053         30.0     5.94  

[28923054 rows x 11 columns]

In [52]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
plt.figure(figsize=(18, 8))
plt.plot('도착시간', '평균 이동 시간(분)', data = movement)
# plt.axhline(movement['이동인구(합)'].mean(), color='r')
plt.xticks(rotation=30)
plt.grid()
plt.show()

OverflowError: Exceeded cell block limit in Agg.  Please set the value of rcParams['agg.path.chunksize'], (currently 0) to be greater than 100 or increase the path simplification threshold(rcParams['path.simplify_threshold'] = 0.111111111111 by default and path.simplify_threshold = 0.111111111111 on the input).

<Figure size 1800x800 with 1 Axes>

* 4월의 요일 수를 알려드릴게요
    * 일 : 4, 월 : 4, 화 : 4, 수 : 4, 목 : 4, 금 : 5, 토 : 5

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.


#### [실습문제5] 어느 요일의 심야 이동이 가장 많을까요?

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 0:00~ 4:59

nightmove = movement[movement['도착시간'] < 5 ]
nightmove['요일'].value_counts()

#### [실습문제6] 서울 지역의 데이터만 골라 보도록 하겠습니다.

* 우리는 서울시의 데이터만 필요합니다!

* 출발과 도착이 모두 서울인 데이터만 추출해 주세요.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# seoul_moving = seoul_moving[(seoul_moving["출발 시군구 코드"] < 20000) & (seoul_moving["도착 시군구 코드"] < 20000) ]
# seoul_moving

onlyseoul = movement[(movement['출발 시군구 코드'] < 20000) & (movement['도착 시군구 코드'] < 20000)]
onlyseoul

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
print(onlyseoul['출발 시군구 코드'].value_counts())

In [ ]:
print(onlyseoul['도착 시군구 코드'].value_counts())

#### [실습문제7] 서울지역의 구 별로 groupby하여 데이터를 확인하겠습니다 (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* 참고자료 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (출발 시군구 코드)
# bus1.groupby('자치구')['노선명'].nunique()

onlyseoul.groupby(by=['출발 시군구 코드'], as_index = False)['이동인구(합)'].sum()


In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (도착 시군구 코드)
onlyseoul.groupby(by=['도착 시군구 코드'], as_index = False)['이동인구(합)'].sum()


#### [실습문제8] 낮시간만 볼게요 심야버스 제외하겠습니다

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seouldaytime = onlyseoul[onlyseoul['도착시간'] >= 5]
seouldaytime.reset_index(drop=True, inplace = True)
seouldaytime

#### [실습문제9] '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)' , '이동인구(합)' 의 평균 과 합을 구해보겠습니다

* groupby 사용, as_index=False로 설정해주세요

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoulmean = seouldaytime.groupby(by=['도착 시군구 코드'], as_index = False)['평균 이동 시간(분)', '이동인구(합)', '도착시간'].mean()
seoulmean

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

seoulsum = seouldaytime.groupby(by=['도착 시군구 코드'], as_index = False)['평균 이동 시간(분)', '이동인구(합)'].sum()
seoulsum

In [ ]:
# 컬럼명을 변경해볼까요?
seoulsum.rename(columns={"평균 이동 시간(분)":'총 이동 시간', '이동인구(합)':'총 이동 인구'},inplace=True)

In [ ]:
seoulsum

In [ ]:
# 파일을 합쳐볼까요
df_seoul_moving = pd.merge(seoulmean, seoulsum, how = 'inner', on = '도착 시군구 코드')

In [ ]:
df_seoul_moving['자치구'] = ['종로구', '중구', '용산구', '성동구', '광진구', '동대문구', '중랑구', '성북구', '강북구', '도봉구', '노원구', '은평구', '서대문구', '마포구', '양천구', '강서구', '구로구', '금천구', '영등포구', '동작구', '관악구', '서초구', '강남구', '송파구', '강동구']
df_seoul_moving

In [ ]:
# 해당 데이터프레임을 'df_seoul_moving.csv' 파일로 저장하세요
df_seoul_moving.to_csv("C:/Users/User/iCloudDrive/AIVLE/3주차/2023.02.13_미니프로젝트1차_실습자료/df_seoul_moving.csv", index = False)

---

# 3.데이터 분석하기
+ KeyPoint : 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별할 수 있다.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# 시각화 한글폰트 설정
# 맑은 고딕

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

# scipy
import scipy.stats as spst

#### [실습문제10] 데이터 분포 알아보기
* 다양한 변수를 기준으로 그래프를 그려보고 인사이트를 도출해보세요.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
df_seoul_moving 

In [ ]:
df_seoul_moving['시군구'] = df_seoul_moving['도착 시군구 코드'].Series.slice(start=0, stop=4)
# df_seoul_moving['자치구'] = df_seoul_moving['시군구'].map({'01':'종로구', '02.0':'중구', '03.0': '용산구', '04.0':'성동구', '05.0': '광진구', '06.0': '동대문구', '07.0': '중랑구', '08.0': '성북구', '09.0': '강북구', '10.0': '도봉구', '11.0': '노원구', '12.0': '은평구', '13.0': '서대문구', '14.0': '마포구', '15.0': '양천구', '16.0': '강서구', '17.0': '구로구', '18.0': '금천구', '19.0': '영등포구', '20.0': '동작구', '21.0': '관악구', '22.0': '서초구', '23.0': '강남구', '24.0': '송파구', '25.0': '강동구'})
df_seoul_moving

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
plt.figure(figsize=(18, 8))
plt.plot('도착 시군구 코드', '이동인구(합)', data = df_seoul_moving)
plt.axhline(df_seoul_moving['이동인구(합)'].mean(), color='r')
plt.xticks(rotation=30)
plt.show()

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

In [ ]:
# 위 차트를 통해 알게된 사실을 정리해봅시다.
# 1. 
# 2.
# 3.

---